### ISSUE TITLE
feat: out-of-band sync for hubs

### ISSUE THREAD

**Issue by varunsrin** [farcasterxyz/hub#66]
Implement a sync protocol that allows two hubs to fully replicate and converge to the same state across all sets

**Comment by sanjayprabhu** [comment]
I've been thinking about this problem. Here's some high level thoughts:

1. All messages have a content based cryptographic hash and a timestamp. They are immutable.
2. There is **no guaranteed ordering** (messages can arrive in any order and it's possible to set message timestamps in the past). 
3. The vast majority of the time, **incoming messages will be ordered by timestamp (roughly)**. e.g. it's highly unlikely we receive a message from a previous year. So, we should optimize for this case the most. 
4. The approach we choose **should support sharding by user** eventually.
5. In addition to the above, there are two other scenarios we should be able to handle well: a) the initial sync (no data) and b) when a hub has been offline for an extended period.
6. Because the sync happens over the network, we need to be mindful of network traffic/amount of data required to determine where and how two hubs diverge.
7. Must **be able to handle light clients** (who may not choose to store all of the actual messages). However, we assume the light clients will at least store the timestamp and the hash.

Because of the above considerations, I believe we need some form of merkle based (because of 1), timestamp ordered (because of 2) datastructure in order to sync efficiently.

I've looked into and discarded the following:
1. **Naïve merkle trees** obviously won't work because there's no way to determine where the change happened (e.g. whether a hub is missing a message from 2012 or 2021).
2. **Merkle b+ trees** with keys ordered by timestamp: Initially seemed promising however, it doesn't work because the key ordering is non-deterministic (depends on insertion order). Which mean you get different merkle roots for the same dataset.
3. **Merkle mountain ranges**: In order to determine where a change occurred, we have to order the data by timestamp. However, this means that if we receive an out of order message we have to reconstruct (potentially) large parts of the datastructure.

I'm currently investigating another option that I think will handle all of our requirements. We could use a modified version of ethereum's Merkle Patricia Trie. The main change would be that we enforce the keys to have the form `"<timestamp><hash>"`. You could also think of it as a trie of tries, first by timestamp and then by message hash. This gives us a bunch of nice properties:

1. Merkle root determines state of the entire tree
2. Messages are ordered by timestamp
3. Inserts of recent messages have high locality, but we can also handle out of order messages with no additional overhead
4. By comparing the the hashes of every node of the timestamp portion of the most recently inserted key, two hubs can quickly determine where in the timeline their state diverges. This is the key benefit.

I'll share a doc later this week outlining the sync algorithm in more detail.

**Comment by varunsrin** [comment]
One other approach might be to use bloom filters with a pull/push model, similar to what [Solana does](https://docs.solana.com/validator/gossip). A node can gossip its current state as a bloom filter (a pull), any recipient can compare the state to its own state and find delta messages, and "push" the new state back. Bloom filters can be constructed on the fly by timestamp range as the set size gets larger.

**Comment by sanjayprabhu** [comment]
Interesting. That's pretty cool. I think conceptually the two approaches are quite similar, creating bloomfilters by timestamp range and having a merkle trie of messages ordered by timestamp serve the same purpose. The main benefit I see with the trie is it automatically handles bucketing of messages by timestamps, and it provides a way to efficiently determine where two hubs diverge. 

e.g. If we have fixed timestamp ranges for the bloomfilter, then we'd need to send over O(number of buckets) to determine where messages diverge if the divergence is not in the most recent range. Further, once we determine which timestamp range the hubs diverge, we'd have to send over all the messages within that range to get the hubs in sync. The trie ordered by timestamp solves this by imposing a hierarchical order. It's possible to determine the lowest level of the trie where hubs diverge in O(1) time.

I'm almost done with the doc, will share tomorrow.

**Comment by sanjayprabhu** [comment]
Here's the doc https://sanjayr.notion.site/Farcaster-Sync-Algorithm-2ffa371e76d64435b31249f8abb57d89

**Comment by sanjayprabhu** [comment]
High level plan for the implementation:

 - [ ] Introduce a merkle trie that keeps track of messages merged by the engine (https://github.com/farcasterxyz/hub/pull/145)
 - [ ] Add RPC calls so two hubs can diff their tries and request missing messages (diff sync is possible at this point)
 - [ ] optimizations: cache the trie state for hub restarts, repair/re-create cache from db
 - [ ] optimizations: faster hash algorithms, MPT extension nodes for the trie
 - [ ] metrics/logging/visibility into sync performance
 - [ ] snapshot sync for bootstrapping a new node

**Comment by varunsrin** [comment]
added a separate ticket to track snapshot sync #155

**Comment by varunsrin** [comment]
@sanjayprabhu one important issue will be defining the boundary between storage, sync and p2p. for instance, assume that a client sends 5 Farcaster messages to a hub and it gossips out 5 Gossip messages with state trie roots. A few questions: 

1. Is it guaranteed that the state trie increments atomically and each Gossip message will have a unique trie root? 
2. Will the receiving nodes merge messages in sequence or parallel? 
3. Will this create unnecessary thrash because we have no ordering of the state? that is, we attempt a full sync after receiving Gossip Message 1 and reach the final state after Gossip message 5. But then we ingest Gossip Message 2 which has a different state root which will trigger another sync because of a root mismatch.

**Comment by sanjayprabhu** [comment]
My current plan is to use a sync buffer of 10 seconds (customizable but same for all the hubs on the network). For every gossip message to be sent, we pick the most recent timestamp that's divisible by 10 in the trie, and used that as the reference timestamp to calculate the excluded hash. Instead of the state root, we'll send the concatenated excluded hash for this timestamp (this also enables hubs to determine where they diverge without an RPC call, at the cost of slightly higher bandwidth). If bandwidth is a problem, we can also just hash this as the excluded root.

In effect, it's as if we're comparing the trie at a rolling, deterministic point in time in the past for all hubs, and ignoring any changes that happen after that point. 

The trie updates can happen in parallel with regular gossip messages, but will generally not affect sync because of the buffer. While a hub is syncing, it's excluded hash will keep changing while the other hubs remain constant. But note that since we'll be sending along the total number of messages a hub has for that timestamp, the hubs with more messages will not choose to sync with the hub with fewer messages (the one that's out of date).

**Comment by varunsrin** [comment]
closing in favor of the new tickets that @adityapk00 is working on
